src/rules/__tests__/test-utils.ts

```diff
@@ -1,6 +1,87 @@
 import { createRequire } from 'module';
+import { TSESLint } from '@typescript-eslint/utils';
+import { version as eslintVersion } from 'eslint/package.json';
+import * as semver from 'semver';
 
 const require = createRequire(__filename);
 const eslintRequire = createRequire(require.resolve('eslint'));
 
 export const espreeParser = eslintRequire.resolve('espree');
+
+export const usingFlatConfig = () => semver.major(eslintVersion) >= 9;
+
+export class FlatCompatRuleTester extends TSESLint.RuleTester {
+  public constructor(testerConfig?: TSESLint.RuleTesterConfig) {
+    super(FlatCompatRuleTester._flatCompat(testerConfig));
+  }
+
+  public override run<
+    TMessageIds extends string,
+    TOptions extends Readonly<unknown[]>,
+  >(
+    ruleName: string,
+    rule: TSESLint.RuleModule<TMessageIds, TOptions>,
+    tests: TSESLint.RunTests<TMessageIds, TOptions>,
+  ) {
+    super.run(ruleName, rule, {
+      valid: tests.valid.map(t => FlatCompatRuleTester._flatCompat(t)),
+      invalid: tests.invalid.map(t => FlatCompatRuleTester._flatCompat(t)),
+    });
+  }
+
+  /* istanbul ignore next */
+  private static _flatCompat<
+    T extends
+      | undefined
+      | TSESLint.RuleTesterConfig
+      | string
+      | TSESLint.ValidTestCase<unknown[]>
+      | TSESLint.InvalidTestCase<string, unknown[]>,
+  >(config: T): T {
+    if (!config || !usingFlatConfig() || typeof config === 'string') {
+      return config;
+    }
+
+    const obj = { languageOptions: { parserOptions: {} } };
+
+    for (const [key, value] of Object.entries(config)) {
+      if (key === 'parser') {
+        // @ts-expect-error this is expected
```

Comment:
what is expected? (applies to all)

Reply:
that it errors, because it's very annoying to type especially given we're still a major version of `@typescript-eslint` behind (so the flat config types are not necessarily correct and properly supported in the utils we're using).

This is effectively a kludge to tie us over the next couple of months until v9 actually comes out, like some of the other casts and `@ts-expect-errors` we've got - I'll try to stick a general explainer comment in but every time I tried it didn't come out that subscint and I'm not expecting it to require much attention from anyone other than me 😅

Reply:
I've refactored this with types which let me remove most of the errors - ironically the last one I could get rid of by using dedicated `if` statements but I don't think it's worth it